tracer/test/Datadog.Trace.Tests/Sampling/SpanSamplerTests.cs

```diff
@@ -0,0 +1,236 @@
+// <copyright file="SpanSamplerTests.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Datadog.Trace.Sampling;
+using Datadog.Trace.Util;
+using FluentAssertions;
+using Xunit;
+
+namespace Datadog.Trace.Tests.Sampling
+{
+    [Collection(nameof(Sampling))]
+    public class SpanSamplerTests
+    {
+        private static readonly ulong Id = 1;
+        private static readonly Span CartCheckoutSpan = new Span(new SpanContext(Id++, Id++, null, serviceName: "shopping-cart-service"), DateTimeOffset.Now) { OperationName = "checkout" };
+
+        [Fact]
+        public void Constructor_ShouldThrow_WhenNullRulesGiven()
+        {
+            var exception = Assert.Throws<ArgumentNullException>(() => new SpanSampler(null));
+            exception.ParamName.Should().Be("rules");
+        }
+
+        [Fact]
+        public void NoMatchingRules_ShouldNot_TagSpan()
+        {
+            var rule1 = new SpanSamplingRule("serviceName", "operationName", 1.0f, 500.0f);
+            var rule2 = new SpanSamplingRule("serviceName2", "operationName2", 1.0f, 500.0f);
+            var rules = new List<SpanSamplingRule>() { rule1, rule2 };
+            var sampler = new SpanSampler(rules);
+            var span = new Span(new SpanContext(5, 6, null, serviceName: "service-name"), DateTimeOffset.Now) { OperationName = "operation-name" };
+
+            sampler.MakeSamplingDecision(span);
+
+            span.Tags.GetTag(Tags.SingleSpanSampling.RuleRate).Should().BeNull();
+            span.Tags.GetTag(Tags.SingleSpanSampling.MaxPerSecond).Should().BeNull();
+            span.Tags.GetTag(Tags.SingleSpanSampling.SamplingMechanism).Should().BeNull();
+        }
+
+        [Fact]
+        public void FirstMatchingRule_Should_Tag()
+        {
+            var expectedRuleRate = 1.0f;
+            var expectedMaxPerSecond = 500.0f;
+            var expectedSamplingMechanism = 8; // SamplingMechanism.SpanSamplingRule
+            var rule1 = new SpanSamplingRule("service-name", "operation-name", 1.0f, 500.0f);
```

Comment:
You might also want to do another test like this where the sample rate for the first rule is 0 and the sample rate for the second rule is 1. We would expect the second rule to always apply

Reply:
Looking at it, yeah that is how it would currently work, but I actually think that I have my implementation incorrect regarding the RFC.

>If a span matches two or more rules, then the rule that appears earlier in the JSON array is the one that applies, i.e. the order in which rules are specified defines a priority.
>
>Rules need be evaluated against every span. If a span's service or operation name can change, then evaluate the rule against the span only after all changes have been committed.

And my logic is currently 

```csharp
foreach (var rule in _rules)
{
    if (rule.IsMatch(span) && rule.ShouldSample(span))
    {
        AddTags(span, rule);
        return true;
    }
}
```

So I wouldn't actually honor the first rule taking priority over the second.

Reply:
So I might need to alter my logic to return `false` when a rule matches, but isn't sampled.

Reply:
I've updated the logic in 2ac58254d1e1ae01a42b07d0876f5ba5a891cc94 to prioritize the first rule over the second

So in this case if the sample rate is 0 for the first rule and the sample rate for the second rule is 1, the _first_ rule would always apply and the span wouldn't be sampled.